File: 2/File copy/c/FileCopy.c
#include<stdio.h>
#include<fcntl.h>
#include<unistd.h> // For ssize_t

#define BUFFER_SIZE 4096
// https://www.quora.com/Why-is-buffer-size-set-in-a-C-program

int main(int argc, char* argv[]) {
	int fin, fout;
	ssize_t ret_in, ret_out; // https://jameshfisher.com/2017/02/22/ssize_t.html
	char buffer[BUFFER_SIZE];

	if(argc != 3) {
		printf("Usage: FileCopy source destination");
		return 0;
	}

	fin = open(argv[1], O_RDONLY); // Read only

	if(fin == -1) {
		perror("Source file open error");
		return 0;
	}

	fout = open(argv[2], O_WRONLY | O_CREAT, 0644); // Write only bitwise OR with create if not exist; give permission to write.
	 // http://pubs.opengroup.org/onlinepubs/9699919799/functions/open.html
	
	if(fout == -1) {
		perror("Destination file opening/creating issue");
		return 0;
	}

	while((ret_in = read(fin, &buffer, BUFFER_SIZE)) > 0) { // http://man7.org/linux/man-pages/man2/read.2.html
		ret_out = write(fout, &buffer, ret_in); // http://man7.org/linux/man-pages/man2/write.2.html
		if(ret_in != ret_out) {
			perror("Error copying");
			return 0;
		}
	}

	close(fin);
	close(fout);

	return 1;
}

File: 2/File copy/java/src/FileCopy.java
import java.io.*;

public class FileCopy {
    public static void main(String[] args) throws IOException{
        int i;
        FileInputStream fin = null;
        FileOutputStream fout = null;
        if(args.length !=  2) {
            System.out.println("Usage: FileCopy source destination");
            return;
        }
        try{
            fin = new FileInputStream(args[0]);
            fout = new FileOutputStream(args[1]);

            do {
                i = fin.read();
                if(i != -1) fout.write(i);
            } while(i != -1);
        } catch (IOException e) {
            System.out.println("I/O Error: " + e);
        } finally {
            try {
                if (fin != null) fin.close();
            } catch (IOException e1) {
                System.out.println("Error closing input file: " + e1);
            }
            try {
                if(fout != null) fout.close();
            } catch (IOException e2) {
                System.out.println("Error closing output file: " + e2);
            }
        }
    }
}
